open-telemetry/opentelemetry-collector-contrib
# [receiver/kubeletstats] Add feature gate for cpu utilization metrics' deprecation

**Description:** <Describe what has changed.>


This PR adds a feature gate as discussed at https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/27885#issuecomment-2285476787.
If the feature gate is enabled the `container.cpu.utilization`, `k8s.pod.cpu.utilization` and `k8s.node.cpu.utilization` metrics will be disabled being replaced by the `container.cpu.usage`, `k8s.pod.cpu.usage` and `k8s.node.cpu.usage`.

### Feature gate schedule

- alpha: when enabled it makes the .cpu.usage metrics enabled by default
- beta: .cpu.usage metrics are enabled by default and any configuration enabling the deprecated .cpu.utilization metrics will be failing. Explicitly disabling the feature gate provides the old (deprecated) behavior.
- stable: .cpu.usage metrics are enabled by default and the deprecated metrics are completely removed.
- Removed three releases after `stable`.

@TylerHelmuth @dmitryax let me know what you think about this approach.

**Link to tracking Issue:** <Issue number if applicable>

**Testing:** <Describe what testing was performed and which tests were added.>

**Documentation:** <Describe the documentation added.>

### How to test this

1. Using the following configuration
```yaml
mode: daemonset
presets:
  kubeletMetrics:
    enabled: true

image:
  repository: otelcontribcol-dev
  tag: "latest"
  pullPolicy: IfNotPresent

command:
  name: otelcontribcol
  extraArgs: [--feature-gates=receiver.kubeletstats.enableCPUUsageMetrics]

config:
  exporters:
    debug:
      verbosity: normal
  receivers:
    kubeletstats:
      collection_interval: 10s
      auth_type: 'serviceAccount'
      endpoint: '${env:K8S_NODE_NAME}:10250'
      insecure_skip_verify: true

  service:
    pipelines:
      metrics:
        receivers: [kubeletstats]
        processors: [batch]
        exporters: [debug]
```
2. Ensure that only the `.cpu.usage` metrics are reported.
3. Disable the feature gate and check that only the `.cpu.utilization` metrics are reported.
